californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java

```diff
@@ -466,8 +489,9 @@ public boolean isComplete() {
 	public void setComplete() {
 		this.complete = true;
 		ExchangeObserver obs = this.observer;
-		if (obs != null)
+		if (obs != null) {
```

Comment:
I don't know, insert such a call after `this.complete = true;` should work on my opinion. 
But though this PR's intention is to make the states visible to other threads , I didn't redesign the interface nor implementation.